### PR TITLE
Make it possible to use different templates

### PR DESCRIPTION
**Pull request by bjorvack**
https://app.activecollab.com/108877/projects/62/tasks/30160

### CODE REVIEW

**Review by ohvitorino — APPROVED** [review]
👍

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 1]
```diff
@@ -80,4 +85,27 @@ private function handleForm()
...
+
+    /**
+     * @return array
+     */
+    private function getPossibleTemplates()
+    {
+        $templates = array();
+        $finder = new Finder();
+        $finder->name('*.html.twig');
+        $finder->in(FRONTEND_MODULES_PATH . '/Slideshows/Layout/Widgets');
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: It's better if you use `$this->getModule()` instead of `Slideshows`

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 2]
```diff
@@ -80,4 +85,27 @@ private function handleForm()
...
+        // if there is a custom theme we should include the templates there also
+        if (Theme::getTheme() != 'core') {
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: This check is pretty useless I think. If someone wants to make a 'core' theme, the extra templates won't be loaded. 

In Fork 5 the default theme will be `Fork` so people should be able to expand on that

File: src/Backend/Modules/Slideshows/Installer/Data/install.sql [hunk 3]
```diff
@@ -4,6 +4,7 @@ CREATE TABLE IF NOT EXISTS `slideshows` (
   `language` varchar(20) COLLATE utf8_unicode_ci NOT NULL,
   `created_on` datetime NOT NULL,
   `extra_id` int(11) NOT NULL,
+  `template` varchar(20) COLLATE utf8_unicode_ci NOT NULL DEFAULT 'Detail.html.twig',
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Installer/Data/install.sql:7]: This should also be added to the test database

File: src/Backend/Modules/Slideshows/Actions/Edit.php [hunk 4]
```diff
@@ -138,4 +147,27 @@ private function handleForm()
             }
         }
     }
+
+    /**
+     * @return array
+     */
+    private function getPossibleTemplates()
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Actions/Edit.php]: This method should be extracted to the model class

File: src/Frontend/Modules/Slideshows/Widgets/Detail.php [hunk 5]
```diff
@@ -50,4 +50,16 @@ private function parse()
         // assign data
         $this->tpl->assign('slideshow', $this->item);
     }
+
+    /**
+     * @return string
+     */
+    private function getSlideShowTemplate()
+    {
+        if (Theme::getTheme() != 'code') {
```
Comment (StijnVrolijk) [inline src/Frontend/Modules/Slideshows/Widgets/Detail.php]: typo in `'core'`

Comment (StijnVrolijk) [reply src/Frontend/Modules/Slideshows/Widgets/Detail.php]: also, wouldn't it be a good idea to check if this template still actually exists?

**Review by StijnVrolijk — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Actions/Edit.php (hunk 4 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Edit.php]: fixed 875b6a3

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 1 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: moved to model
fixed 875b6a3

File: src/Backend/Modules/Slideshows/Installer/Data/install.sql (hunk 3 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Installer/Data/install.sql:7]: @StijnVrolijk there doesn't seem to be a test database

File: src/Frontend/Modules/Slideshows/Widgets/Detail.php (hunk 5 above)
Comment (bjorvack) [reply src/Frontend/Modules/Slideshows/Widgets/Detail.php]: fixed 4a85f3d

File: src/Backend/Modules/Slideshows/Installer/Data/install.sql (hunk 3 above)
Comment (StijnVrolijk) [reply src/Backend/Modules/Slideshows/Installer/Data/install.sql:7]: ah, nvm.. i thought we were in sumocoders/forkcms

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 2 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: @StijnVrolijk and what happens if there is no custom theme?

Comment (StijnVrolijk) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: @bjorvack  no templates will be found because you check if the dir exists 2 lines after this one. If the theme is 'core', the path will be `src/Frontend/Themes/core/Modules/Slideshows/Layout/Widgets` and this path does not exists, therefor finder won't go looking for templates in this path

File: src/Frontend/Modules/Slideshows/Widgets/Detail.php [hunk 6]
```diff
@@ -56,10 +56,11 @@ private function parse()
...
     {
-        if (Theme::getTheme() != 'code') {
-            return FRONTEND_PATH . '/Themes/' . Theme::getTheme() . '/Modules/Slideshows/Layout/Widgets/' . $this->item['template'];
+        $filepath = FRONTEND_PATH . '/Themes/' . Theme::getTheme() . '/Modules/' . $this->getModule() . '/Layout/Widgets/' . $this->item['template'];
+        if (Theme::getTheme() != 'core' && file_exists($filepath)) {
+            return $filepath;
         }
 
-        return FRONTEND_MODULES_PATH . '/Slideshows/Layout/Widgets/' . $this->item['template'];
+        return FRONTEND_MODULES_PATH . '/' . $this->getModule() . '/Layout/Widgets/' . $this->item['template'];
```
Comment (StijnVrolijk) [inline src/Frontend/Modules/Slideshows/Widgets/Detail.php]: You should also check if this file still exists

**Review by StijnVrolijk — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 namespace Backend\Modules\Slideshows\Actions;
 
+use Symfony\Component\Finder\Finder;
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: do you use these 2 use statements in this class?
They have been added but I don't see a use in the added lines

File: src/Backend/Modules/Slideshows/Actions/Edit.php [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 namespace Backend\Modules\Slideshows\Actions;
 
+use Symfony\Component\Finder\Finder;
+use Frontend\Core\Engine\Theme;
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Actions/Edit.php]: do you use these 2 use statements in this class?
They have been added but I don't see a use in the added lines

File: src/Backend/Modules/Slideshows/Engine/Model.php [hunk 9]
```diff
@@ -227,4 +230,27 @@ public static function getNextSlideSequence($slideshowId)
...
+     * @return array
+     */
+    public static function getPossibleTemplates()
+    {
+        $templates = array();
+        $finder = new Finder();
+        $finder->name('*.html.twig');
+        $finder->in(FRONTEND_MODULES_PATH . '/' . self::MODULE . '/Layout/Widgets');
+        // if there is a custom theme we should include the templates there also
+        if (Theme::getTheme() != 'core') {
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Engine/Model.php]: I think the theme can also be `Core` so maybe make this check case insensitive?

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 10]
```diff
@@ -33,6 +35,8 @@ private function handleForm()
         // create elements
         $txtTitle = $this->frm->addText('title', null, null, 'form-control title', 'form-control danger title');
 
+        $template = $this->frm->addDropdown('template', Model::getPossibleTemplates());
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: you should only show/add this dropdown if there are more than 1 possible template, just like we do in the Content Blocks module

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 11]
```diff
@@ -46,6 +50,7 @@ private function handleForm()
                 // build item
                 $item['language'] = Language::getWorkingLanguage();
                 $item['title'] = $txtTitle->getValue();
+                $item['template'] = $template->getValue();
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: you should only show/add this dropdown if there are more than 1 possible template, just like we do in the Content Blocks module else you should fallback to the default

File: src/Backend/Modules/Slideshows/Actions/Edit.php [hunk 12]
```diff
@@ -103,6 +105,12 @@ private function handleForm()
             'form-control danger title'
         );
 
+        $template = $this->frm->addDropdown(
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Actions/Edit.php]: you should only show/add this dropdown if there are more than 1 possible template, just like we do in the Content Blocks module

File: src/Backend/Modules/Slideshows/Engine/Model.php [hunk 13]
```diff
@@ -23,6 +24,8 @@ class Model
 
     const IMAGE_FOLDER = '/Slideshows/';
 
+    const MODULE = 'Slideshows';
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Engine/Model.php]: why did you move this to a constant? it is the Slideshows module, I don't really see what the use is of having it is a constant

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig [hunk 14]
```diff
@@ -16,6 +16,13 @@
         </label>
         {% form_field title %} {% form_field_error title %}
       </div>
+      <div class="form-group">
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: you should only show/add this dropdown if there are more than 1 possible template, just like we do in the Content Blocks module

File: src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig [hunk 15]
```diff
@@ -20,6 +20,15 @@
         </label>
         {% form_field title %} {% form_field_error title %}
       </div>
+
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig:23]: you should only show/add this dropdown if there are more than 1 possible template, just like we do in the Content Blocks module

**Review by carakas — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 2 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: fixed 20c93f3

File: src/Frontend/Modules/Slideshows/Widgets/Detail.php (hunk 6 above)
Comment (bjorvack) [reply src/Frontend/Modules/Slideshows/Widgets/Detail.php]: fixed 34b2e8d

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 7 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: fixed 33bfbed

File: src/Backend/Modules/Slideshows/Actions/Edit.php (hunk 8 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Edit.php]: fixed 33bfbed

File: src/Backend/Modules/Slideshows/Engine/Model.php (hunk 9 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Engine/Model.php]: no longer valid in 34b2e8d

File: src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig (hunk 15 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig:23]: fixed deb4ab9

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig (hunk 14 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: fixed deb4ab9

File: src/Backend/Modules/Slideshows/Actions/Edit.php (hunk 12 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Edit.php]: fixed deb4ab9

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 11 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: fixed deb4ab9

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 10 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: fixed deb4ab9

File: src/Backend/Modules/Slideshows/Engine/Model.php (hunk 13 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Engine/Model.php]: fixed c776f32

**Comment by bjorvack** [comment]
@StijnVrolijk @carakas I made changes based on your reviews.

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig [hunk 16]
```diff
@@ -16,13 +16,15 @@
...
         {% form_field title %} {% form_field_error title %}
       </div>
-      <div class="form-group">
-        <label for="title" class="control-label">
-          {{ 'lbl.Template'|trans|capitalize }}
-          <abbr data-toggle="tooltip" title="{{ 'lbl.RequiredField'|trans|capitalize }}">*</abbr>
-        </label>
-        {% form_field template %} {% form_field_error template %}
-      </div>
+      {% if template %}
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: did you test this?
cause if I remember correctly this won't work since it needs to be ddmTemplate

**Review by carakas — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig [hunk 17]
```diff
@@ -16,6 +16,15 @@
         </label>
         {% form_field title %} {% form_field_error title %}
       </div>
+      {% if template %}
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: did you test this? I think it should be ddmTemplate

File: src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig [hunk 18]
```diff
@@ -20,6 +20,17 @@
...
+      {% if template %}
```
Comment (carakas) [inline src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig]: did you test this? I think it should be ddmTemplate

**Review by carakas — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig (hunk 17 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: fixed 9617a04

File: src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig (hunk 18 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Layout/Templates/Edit.html.twig]: fixed 9617a04

File: src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig (hunk 16 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Layout/Templates/Add.html.twig]: fixed 9617a04

**Review by carakas — APPROVED** [review]

File: src/Backend/Modules/Slideshows/Actions/Add.php [hunk 19]
```diff
@@ -33,6 +33,15 @@ private function handleForm()
         // create elements
         $txtTitle = $this->frm->addText('title', null, null, 'form-control title', 'form-control danger title');
 
+        $possibleTemplates = Model::getPossibleTemplates();
+        if (count($possibleTemplates) > 1) {
+            $template = $this->frm->addDropdown(
+                'template',
+                Model::getPossibleTemplates(),
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Actions/Add.php]: you can use `$possibleTemplates` here

File: src/Backend/Modules/Slideshows/Actions/Edit.php [hunk 20]
```diff
@@ -103,6 +103,15 @@ private function handleForm()
             'form-control danger title'
         );
 
+        $possibleTemplates = Model::getPossibleTemplates();
+        if (count($possibleTemplates) > 1) {
+            $template = $this->frm->addDropdown(
+                'template',
+                Model::getPossibleTemplates(),
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Slideshows/Actions/Edit.php]: you can use `$possibleTemplates` here

**Review by StijnVrolijk — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Slideshows/Actions/Edit.php (hunk 20 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Edit.php]: fixed 8cd340a

File: src/Backend/Modules/Slideshows/Actions/Add.php (hunk 19 above)
Comment (bjorvack) [reply src/Backend/Modules/Slideshows/Actions/Add.php]: fixed 8cd340a
